### ISSUE TITLE
Non null terminated strings in em_asm section

### ISSUE THREAD

**Issue by KochetovKirill** [emscripten-core/emscripten#19197]
It appeared that `get_section_strings` in `extract_metadata.py` can't handle strings that don't end with \00.

The story is following. We have our own language in which strings are not null terminated and store their length instead. So we have met some problems when upgrading to a newer version of emscripten.

There was a fix written that we use locally. And i would like to ask if it is possible to merge that fix and, if yes, how to do it better.

**Comment by sbc100** [comment]
`EM_ASM` and `EM_JS` strings are defined as being null terminated (at least right now they are).     Can you explain a little bit more about your setup?   How are you implementing the EM_ASM?  Are you using the C macro provided by emscripten?   Or are you using some kind of special macro in your own language?    Given that those strings have special handling (i.e. they are placed in a custom section) can you also kind some way to make them normal null terminated C strings?

**Comment by sbc100** [comment]
How are you linking your language code with emscripten?   Are you generated `wasm32-unknown-emscripten` object files from your language use your own llvm frontend and then the common llvm backend?

**Comment by KochetovKirill** [comment]
> How are you linking your language code with emscripten? Are you generated `wasm32-unknown-emscripten` object files from your language use your own llvm frontend and then the common llvm backend?

Yep, pretty much like that. Our language compiles into .ll and then all other things are done by llvm.

> How are you implementing the EM_ASM? Are you using the C macro provided by emscripten? Or are you using some kind of special macro in your own language?

In the case of `EM_ASM`, we use our own analog, which consists of some argument preparation (like putting string into em_asm section) and calling `emscripten_asm_const_int`(or `emscripten_asm_const_double`).

> Given that those strings have special handling (i.e. they are placed in a custom section) can you also find some way to make them normal null terminated C strings?

No, we do not have any way to create strings that don't store their length inside our language. For example in .ll our strings look like that: `@string.120 = private unnamed_addr constant {i32, [10 x i8]} {i32 10, [10 x i8] c"someString"}`.


Maybe it would be better if I make a pull request with changes so we can discuss it more substantively.

**Comment by sbc100** [comment]
Sure, feel free to send a PR.  It seems unfortunate that you can't make your toolchain output the format the emscripten expects but if its really possible we can look at other options.  We might have to write the unit tests in assembly but its not impossible it imagine doing that.
